dastardcommander/configure_level_triggers.py

```diff
@@ -0,0 +1,174 @@
+        if not self.zeroTriggers():
+            self.cursor.insertText("X  Failed: no channels known.\n")
+            return
+
+        self.cursor.insertText("2) Turning on 50 ms autotriggers.\n")
+        if not self.autoTriggers():
+            self.cursor.insertText("X  Failed: no channels known.\n")
+            return
+
+        # 3) Collect baseline level data
+        self.cursor.insertText("3) Collecting baseline data (may take several seconds).\n")
+        self.launchRecordMonitor()
+
+    def launchRecordMonitor(self):
+        self.channels_seen = {
+            id:BaselineFinder(self.recordsPerChan) for id in self.dcom.channelIndicesSignalOnly()
+        }
+        self.nchanIncomplete = len(self.channels_seen)
+        self.progressBar.setMaximum(self.recordsPerChan*self.nchanIncomplete)
+        self.zmqthread = QtCore.QThread()
+        self.zmqlistener = status_monitor.ZMQListener(self.dcom.host, 1+self.dcom.port)
+        self.zmqlistener.pulserecord.connect(self.updateReceived)
+
+        self.zmqlistener.moveToThread(self.zmqthread)
+        self.zmqthread.started.connect(self.zmqlistener.data_monitor_loop)
+        QtCore.QTimer.singleShot(0, self.zmqthread.start)
+
+    @pyqtSlot()
+    def finishConfiguration(self):
+        # 4) Return triggers to previous state (maybe zero them first?)
+        self.cursor.insertText("4) Done with baseline data.  Stopping all triggers.\n")
+        self.zeroTriggers()
+        time.sleep(0.5)
+
+        self.cursor.insertText("5) Sending all level triggers\n")
+        positive = self.positivePulseButton.isChecked()
+        threshold = self.levelSpinBox.value()
+        for idx,blf in self.channels_seen.items():
+            level = int(0.5 + blf.baseline() + threshold)
+            ts = {
+                "ChannelIndices": [idx],
+                "AutoTrigger": False,
+                "EdgeTrigger": False,
+                "LevelTrigger": True,
+                "LevelRising": positive,
+                "LevelLevel": level,
+                }
+            self.dcom.client.call("SourceControl.ConfigureTriggers", ts)
+            time.sleep(0.01)
+        self.cursor.insertText("Done! You may close this window.\n")
+
+    @pyqtSlot()
+    def done(self, dialogCode):
+        """Cleanly close the zmqlistener before closing the dialog."""
+        if self.zmqlistener is not None:
+            self.zmqlistener.running = False
+        if self.zmqthread is not None:
+            self.zmqthread.quit()
+            self.zmqthread.wait()
+        super().done(dialogCode)
+
+    def zeroTriggers(self):
+        ids = self.dcom.channelIndicesAll()
+        if len(ids) == 0:
+            return False
+        config = {
+            "ChannelIndices": ids,
+        }
+        self.dcom.client.call("SourceControl.ConfigureTriggers", config)
+        return True
+
+    def autoTriggers(self):
+        ids = self.dcom.channelIndicesAll()
```

Comment:
should we respect the off channels here? Do you want to re-use the noise trigger setter from the triggering tab?

Reply:
Good point. I was definitely going to respect the blocked channels, and then I think I got distracted.

I don't intend to merge this until we have a chance to try it with real x rays on a real spectrometer next week, so that gives a little time for me to figure that out.